extension-workspace-reference.md

```diff
@@ -0,0 +1,111 @@
+
+The workspace reference request is sent from the client to the server to export project-wide external references. The response strictly includes only the locations in the workspace at which symbols _outside_ of the workspace are referenced. That is:
+
+- Excluding any `ReferenceInformation.URI` which is located within the workspace.
+- Excluding any `ReferenceInformation.Name` which is unexported or private (i.e. according to language semantics).
+- Excluding any `ReferenceInformation.Location` which is located in vendored code (e.g. `vendor/...` for Go, `node_modules/...` for JS, .tgz NPM packages, or .jar files for Java).
+
+
+_Request_
+* method: 'workspace/reference'
+* params: `WorkspaceReferenceParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace Reference Request.
+ */
+interface WorkspaceReferenceParams {
+	/**
+	 * A limit on the number of returned results.
+	 */
+	limit: number
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+	/**
+	 * The location in the workspace at which the `symbol` has been referenced.
+	 */
+	location: Location;
+
+	/**
+	 * The name of the symbol that is being referenced. For example:
+	 *
+	 * 	(Go) "ServeHTTP"
+	 *  (JS) "render"
+	 *
+	 */
+	name?: string;
+
+	/**
+	 * The name of the symbol containing the symbol that is being referenced.
+	 * For example:
+	 *
+	 * 	(Go) "Router"
+	 * 	(JS) "ReactMount"
+	 *
+	 */
+	containerName?: string;
+
+	/**
+	 * The file URI of the symbol that is being referenced. It may be either
+	 * a URI to the directory that contains the symbol OR the file that
+	 * contains the symbol, so long as the tuple (name, containerName, uri)
+	 * uniquely identifies the symbol.
+	 *
+	 * If `name` and `containerName` are not present, it indicates that this
+	 * URI location has been referenced but not a particular symbol. For
+	 * example, a Go import statement or a JS `require()` call, etc. which
+	 * references ('imports' or 'includes') a dependency but not a specific
+	 * symbol.
+	 */
+	uri: string;
+}
+```
+* error: code and message set in case an exception happens during the workspace reference request.
+
+## workspace/symbol query filters
+
+As part of this extension, we also require that language servers support two `workspace/symbol` query filters (specified via the `WorkspaceSymbolParams.query` field). These filters are:
+
+- `is:exported`
+  - When this filter is present in the query, only 'exported' symbols (according to language semantics) should be returned.
+- `dir:<directory>`
+  - When this filter is present in the query, only symbols that are located inside the specified directory path should be returned. Language servers should use this to optimize response times by limiting the work they must perform to only the specified directory.
```

Comment:
> The client always knows the workspace root path, so it is always able to construct an absolute URI.

That is not correct unfortunately - `rootPath` is just a path and therefor does not include a protocol, host or port. So if a client works over FTP on another host, there is no way for the server to build a URI from the root path because it doesn't know the host address, only the remote path. We can solve by allowing relative URIs (relative to the "root URI" that is currently not exposed to the server) and/or with https://github.com/Microsoft/language-server-protocol/issues/113